controller/users.go

```diff
@@ -257,16 +258,85 @@ func (c *UsersController) Update(ctx *app.UpdateUsersContext) error {
 func (c *UsersController) List(ctx *app.ListUsersContext) error {
 	return application.Transactional(c.db, func(appl application.Application) error {
 		var err error
+		var identities []*account.Identity
 		var users []*account.User
 		var result *app.UserArray
-		users, err = appl.Users().List(ctx.Context)
-		if err == nil {
+		identityFilters := []func(*gorm.DB) *gorm.DB{}
+		userFilters := []func(*gorm.DB) *gorm.DB{}
+
+		var appIdentities []*app.IdentityData
+
+		/*
+			There are 2 database tables we fetch the data from : identities , users
+			First, we filter on the attributes of identities table - providerType , username
+			After that we use the above result to cummulatively filter on users  - email , company
+		*/
+
+		/*** Start filtering on Identities table ****/
+
+		if ctx.FilterUsername != nil {
+			identityFilters = append(identityFilters, account.IdentityFilterByUsername(*ctx.FilterUsername))
+		}
+		if ctx.FilterRegistrationCompleted != nil {
+			identityFilters = append(identityFilters, account.IdentityFilterByRegistrationCompleted(*ctx.FilterRegistrationCompleted))
+		}
+		// Add more filters when needed , here. ..
+
+		if len(identityFilters) != 0 {
+			identityFilters = append(identityFilters, account.IdentityFilterByProviderType(account.KeycloakIDP))
+			identityFilters = append(identityFilters, account.IdentityWithUser())
+
+			// From a data model perspective, we are querying by identity ( and not user )
+			identities, err = appl.Identities().Query(identityFilters...)
+
+			if err != nil {
+				// TODO: More elaborate error.
+				return jsonapi.JSONErrorResponse(ctx, errors.Wrap(err, "error listing users"))
+			}
+
+			// cumulatively filter out those not matcing the user-based filters.
+			for _, identity := range identities {
+				// this is where you keep trying all other filters one by one for 'user' fields like email.
+				if ctx.FilterEmail == nil || (ctx.FilterEmail != nil && identity.User.Email == *ctx.FilterEmail) {
```

Comment:
could be just:
`if ctx.FilterEmail == nil || identity.User.Email == *ctx.FilterEmail {`